controllers/tekton/pipeline/pipeline_controller.go

```diff
@@ -121,45 +113,87 @@ func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
 	var tknTaskName string
 	for _, taskSpec := range pipeline.Spec.Tasks {
 		tknTaskName = pipeline.Name + "-" + taskSpec.Name
 		if _, err := r.TknClient.TektonV1beta1().
 			Tasks(pipeline.Namespace).
 			Get(ctx, tknTaskName, metav1.GetOptions{}); err != nil {
-			klog.Infof("Tekton Task [%s] was not found in namespace %s.", tknTaskName, pipeline.Namespace)
-			// Tekton Task resource does not exist.
-			// We do nothing here.
+			klog.Infof("unable to find Tekton Task [%s] in namespace %s", tknTaskName, pipeline.Namespace)
+			// Tekton Task resource does not exist, which means we should do nothing here.
 			continue
 		}
 
-		// Tekton Task exists, we need to delete it.
+		// Since Tekton Task exists, we need to delete it.
 		if err := r.TknClient.TektonV1beta1().
 			Tasks(pipeline.Namespace).
 			Delete(ctx, tknTaskName, metav1.DeleteOptions{}); err != nil {
-			// Failed to delete tekton pipelinerun
-			klog.Errorf("Failed to delete Tekton Task [%s] using tekton client.", tknTaskName)
+			klog.Errorf("unable to delete Tekton Task [%s] using tekton client", tknTaskName)
 			return err
 		}
 	}
 
-	// Secondly, we should find and delete all the related Tekton Pipelines
+	// Secondly, we should find and delete all the related Tekton Pipelines.
 	tknPipelineName := pipeline.Spec.Name
 	if _, err := r.TknClient.TektonV1beta1().
 		Pipelines(pipeline.Namespace).
 		Get(ctx, tknPipelineName, metav1.GetOptions{}); err != nil {
-		klog.Infof("Tekton Pipeline [%s] was not found in namespace %s.", tknPipelineName, pipeline.Namespace)
-		// Tekton Pipeline resource does not exist.
-		// We do nothing here.
-	} else {
-		// If there exists related Tekton Pipelines,
-		// we will delete it.
+		// Tekton Pipeline resource does not exist and that means we ought to do nothing here.
+		klog.Infof("unable to find Tekton Pipeline [%s] in namespace %s", tknPipelineName, pipeline.Namespace)
+		return nil
+	}
+
+	// If there exists related Tekton Pipelines, we will delete it.
+	if err := r.TknClient.TektonV1beta1().
+		Pipelines(pipeline.Namespace).
+		Delete(ctx, tknPipelineName, metav1.DeleteOptions{}); err != nil {
+		// When we failed to delete Tekton PipelineRun resource, then we should return an error here.
+		klog.Errorf("unable to delete Tekton Pipeline [%s] using tekton client", tknPipelineName)
+		return err
+	}
+
+	// What's more, there is no need to hold related PipelineRun CRD resources if
+	// the Pipeline CRD is deleted.
+	// As a result, we need to delete all the devops and Tekton PipelineRun CRD resources,
+	// whose pipelineRef is the deleting pipeline, when deleting a Pipeline.
+
+	// 1. Clean devops PipelineRun CRD resources
+	devopsPipelineRunList := devopsv2alpha1.PipelineRunList{}
+	if err := r.List(ctx, &devopsPipelineRunList); err != nil {
+		klog.Error(err, "unable to fetch pipeline crd resources")
+		return err
+	}
+	// delete the relevant devops PipelineRun resources
+	for _, devopsPipelineRun := range devopsPipelineRunList.Items {
```

Comment:
Okay, I will try to figure it out.

Reply:
> Please consider to use the following method to delete a collection:
> 
> `DeleteAllOf(ctx context.Context, obj runtime.Object, opts ...DeleteAllOfOption) error`
> 
> I'm not sure if it can work well. But it worth to figure it out.

In the function `DeleteAllOf`, we can only use Label Selector or Field Selector to find all matched resources, e.g. see the example [here](https://github.com/kubernetes-sigs/controller-runtime/blob/386c2b5b29bace1e1e0b35afd0f6c1acc38cd140/pkg/client/example_test.go#L234).

Labels or Field can be added, but this operation will be set in the crd yaml file, which may increase user burden. So I use for loop to delete resources here.